### PR TITLE
Add external access shims for TypeScript editor features layer.

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@sharwell @jasonmalinowski @amcasey @jinujoseph

**Comment by tmat** [comment]
@jasonmalinowski @sharwell Ping

File: src/NuGet/VisualStudio/VS.ExternalAPIs.Roslyn.Package.csproj [hunk 1]
```diff
@@ -87,14 +82,8 @@
       <_File Include="$(ArtifactsBinDir)Microsoft.CodeAnalysis.EditorFeatures.Wpf\$(Configuration)\net472\Microsoft.CodeAnalysis.EditorFeatures.Wpf.dll" TargetDir="" />
       <_File Include="$(ArtifactsBinDir)Microsoft.CodeAnalysis.EditorFeatures\$(Configuration)\net472\Microsoft.CodeAnalysis.EditorFeatures.dll" TargetDir="" />
       <_File Include="$(ArtifactsBinDir)Microsoft.CodeAnalysis.ExternalAccess.Apex\$(Configuration)\net472\Microsoft.CodeAnalysis.ExternalAccess.Apex.dll" TargetDir="" />
-      <_File Include="$(ArtifactsBinDir)Microsoft.CodeAnalysis.ExternalAccess.CodeLens\$(Configuration)\net472\Microsoft.CodeAnalysis.ExternalAccess.CodeLens.dll" TargetDir="" />
       <_File Include="$(ArtifactsBinDir)Microsoft.CodeAnalysis.ExternalAccess.Debugger\$(Configuration)\net472\Microsoft.CodeAnalysis.ExternalAccess.Debugger.dll" TargetDir="" />
-      <_File Include="$(ArtifactsBinDir)Microsoft.CodeAnalysis.ExternalAccess.FSharp\$(Configuration)\net472\Microsoft.CodeAnalysis.ExternalAccess.FSharp.dll" TargetDir="" />
```
Comment (sharwell) [inline src/NuGet/VisualStudio/VS.ExternalAPIs.Roslyn.Package.csproj]: ❓ Did you mean to remove this one?

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptSignatureHelpClassifierFactory.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Text;
+using Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Presentation;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript
+{
+    internal static class VSTypeScriptSignatureHelpClassifierFactory
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptSignatureHelpClassifierFactory.cs]: ❔ Why is this needed? It's not exposed to partner code so the implementation should be able to directly call the target method.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugLocationInfo.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api
+{
+    internal struct VSTypeScriptDebugLocationInfo
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugLocationInfo.cs]:
```suggestion
    internal readonly struct VSTypeScriptDebugLocationInfoWrapper
```

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugLocationInfo.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        public readonly int LineOffset;
+
+        public VSTypeScriptDebugLocationInfo(string name, int lineOffset)
+        {
+            Debug.Assert(name != null);
+            Name = name;
+            LineOffset = lineOffset;
+        }
+
+        internal bool IsDefault => Name == null;
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugLocationInfo.cs]: This should be updated to use the wrapper pattern.

https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/ExternalAccess/UnitTesting/Api/UnitTestingInvocationReasonsWrapper.cs

If the type doesn't need to be _constructed_ by the partner then it only needs the one constructor.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugDataTipInfo.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api
+{
+    internal struct VSTypeScriptDebugDataTipInfo
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugDataTipInfo.cs]:
```suggestion
    internal readonly struct VSTypeScriptDebugDataTipInfoWrapper
```

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugDataTipInfo.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        public VSTypeScriptDebugDataTipInfo(TextSpan span, string text)
+        {
+            Span = span;
+            Text = text;
+        }
+
+        public bool IsDefault
+        {
+            get { return Span.Length == 0 && Span.Start == 0 && Text == null; }
+        }
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugDataTipInfo.cs]: This should be updated to follow the wrapper pattern (see other comment for link to example).

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptBreakpointResolutionResult.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api
+{
+    internal sealed class VSTypeScriptBreakpointResolutionResult
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptBreakpointResolutionResult.cs]:
```suggestion
    internal readonly struct VSTypeScriptBreakpointResolutionResultWrapper
```

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptBreakpointResolutionResult.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        public string LocationNameOpt { get; }
+        public bool IsLineBreakpoint { get; }
+
+        private VSTypeScriptBreakpointResolutionResult(Document document, TextSpan textSpan, string locationNameOpt, bool isLineBreakpoint)
+        {
+            Document = document;
+            TextSpan = textSpan;
+            LocationNameOpt = locationNameOpt;
+            IsLineBreakpoint = isLineBreakpoint;
+        }
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptBreakpointResolutionResult.cs]: 📝 This should be updated to follow the wrapper type pattern (see link to example in other comment).

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptLanguageDebugInfoService.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api
+{
+    internal interface IVSTypeScriptLanguageDebugInfoService
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptLanguageDebugInfoService.cs]: The name should be updated to distinguish implementation interfaces (implementation lives in partner code) from accessor interfaces (implementation lives in roslyn internal code).

```suggestion
    internal interface IVSTypeScriptLanguageDebugInfoServiceImplementation
```

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptBreakpointResolutionService.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api
+{
+    internal interface IVSTypeScriptBreakpointResolutionService
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptBreakpointResolutionService.cs]:
```suggestion
    internal interface IVSTypeScriptBreakpointResolutionServiceImplementation
```

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptBreakpointResolutionService.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            var result = await _service.ResolveBreakpointAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (result == null)
+            {
+                return null;
+            }
+
+            return result.IsLineBreakpoint ?
+                BreakpointResolutionResult.CreateLineResult(result.Document, result.LocationNameOpt) :
+                BreakpointResolutionResult.CreateSpanResult(result.Document, result.TextSpan, result.LocationNameOpt);
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptBreakpointResolutionService.cs]: After updating the wrappers to follow the wrapper pattern, this should just be:

```
return result.UnderlyingObject;
```

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptBreakpointResolutionService.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        public Task<IEnumerable<BreakpointResolutionResult>> ResolveBreakpointsAsync(Solution solution, string name, CancellationToken cancellationToken = default)
+            => Task.FromResult(SpecializedCollections.EmptyEnumerable<BreakpointResolutionResult>());
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptBreakpointResolutionService.cs]: 📝 This should delegate to `_service`

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptLanguageDebugInfoService.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VSTypeScriptLanguageDebugInfoService(IVSTypeScriptLanguageDebugInfoService service)
+        {
+            _service = service;
+        }
+
+        public async Task<DebugDataTipInfo> GetDataTipInfoAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var result = await _service.GetDataTipInfoAsync(document, position, cancellationToken).ConfigureAwait(false);
+            return new DebugDataTipInfo(result.Span, result.Text);
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptLanguageDebugInfoService.cs]:
```suggestion
            return result.UnderlyingObject;
```

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptLanguageDebugInfoService.cs [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        }
+
+        public async Task<DebugLocationInfo> GetLocationInfoAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var result = await _service.GetLocationInfoAsync(document, position, cancellationToken).ConfigureAwait(false);
+            return result.IsDefault ? default : new DebugLocationInfo(result.Name, result.LineOffset);
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptLanguageDebugInfoService.cs]:
```suggestion
            return result.UnderlyingObject;
```

File: src/EditorFeatures/CSharp/BlockCommentEditing/BlockCommentEditingCommandHandler.cs [hunk 15]
```diff
@@ -1,35 +1,86 @@
...
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Editor.CSharp.BlockCommentEditing
 {
     [Export(typeof(ICommandHandler))]
     [ContentType(ContentTypeNames.CSharpContentType)]
     [Name(nameof(BlockCommentEditingCommandHandler))]
     [Order(After = PredefinedCompletionNames.CompletionCommandHandler)]
-    internal class BlockCommentEditingCommandHandler : AbstractBlockCommentEditingCommandHandler
+    internal sealed class BlockCommentEditingCommandHandler : ICommandHandler<ReturnKeyCommandArgs>
```
Comment (sharwell) [inline src/EditorFeatures/CSharp/BlockCommentEditing/BlockCommentEditingCommandHandler.cs]: ❓ Why did this command handler change in this pull request?

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugLocationInfo.cs (hunk 4 above)
Comment (tmat) [reply src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugLocationInfo.cs]: I just copied F# implementation of the same:
https://github.com/dotnet/roslyn/blob/af9d30a89b5aeed375f1c4f8f459eb7ac7a74359/src/Tools/ExternalAccess/FSharp/Editor/Implementation/Debugging/IFSharpLanguageDebugInfoService.cs#L10

Comment (tmat) [reply src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugLocationInfo.cs]: The type needs to be constructed. 

I'm not seeing the benefits of the wrapper pattern. Doesn't `UnderlyingObject` expose the type that they shouldn't access?

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugDataTipInfo.cs (hunk 5 above)
Comment (tmat) [reply src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugDataTipInfo.cs]: Why do we need `-Wrapper` suffix when we already have `VSTypeScript-` prefix. Seems like unnecessary clutter.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugLocationInfo.cs (hunk 4 above)
Comment (sharwell) [reply src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptDebugLocationInfo.cs]: > I'm not seeing the benefits of the wrapper pattern. 

Helps link related code together, and minimize duplication.

> Doesn't UnderlyingObject expose the type that they shouldn't access?

No. The analyzer will not allow the property to be used in partner code so it's effectively a proper `internal` visibility.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptSignatureHelpClassifierFactory.cs [hunk 16]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Text;
+using Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Presentation;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api
```
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptSignatureHelpClassifierFactory.cs]: :memo: This file is in the wrong folder.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptLanguageDebugInfoService.cs (hunk 9 above)
Comment (tmat) [reply src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptLanguageDebugInfoService.cs]: Aren't these already distinguished by the `IVSTypeScript` prefix?

**Comment by tmat** [comment]
> Accepting with the statement that the names of both the wrapper types and implementation interfaces 1) do not follow convention and 2) cannot be changed in the future because that would break partner code. Merging in the current state would be a permanent deviation from the expected pattern.

I have updated the PR with a commit that implements the wrapper pattern. I have kept the naming convention used by the F# ExternalAccess implementation. We already have inconsistency between UnitTest EA and F# EA. I don't see value in making the names even longer then they already are by adding `Wrapper` and `Implementation` suffixes. These types are already prefixed with `VSTypeScript` or `FSharp`.

**Comment by tmat** [comment]
@sharwell 
Actually, I see why the `-Implementation` suffix makes sense now. It's common to have `XxxService` implement `IXxxService`, but `FSharpBreakpointResolutionService` does not implement `IFSharpBreakpointResolutionService`. It implements `IBreakpointResolutionService`. So using `IFSharpBreakpointResolutionServiceImplementation` name avoids confusion. I'll rename the TS interfaces.

**Comment by tmat** [comment]
Adding `-Wrapper` makes sense when the wrapping type is a struct but the wrapped type is a reference type. Since then it's more obvious there is some wrapping going on. Since we already will be inconsistent between FSharp and TS in `-Implementation` suffix, I can as well name use the `-Wrapper` suffix as well.
